stdlib/public/core/Result.swift

```diff
@@ -0,0 +1,80 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_frozen
+public enum Result<Value, Error> {
+    case success(Value)
+    case failure(Error)
+    
+    public var value: Value? {
+        switch self {
+        case let .success(value): return value
+        case .failure: return nil
+        }
+    }
+    
+    public var error: Error? {
+        switch self {
+        case let .failure(error): return error
+        case .success: return nil
+        }
+    }
+    
+    public var isSuccess: Bool {
+        switch self {
+        case .success: return true
+        case .failure: return false
+        }
+    }
+    
+}
+
+extension Result where Error: Swift.Error {
+    public func unwrap() throws -> Value {
+        switch self {
+        case let .success(value): return value
+        case let .failure(error): throw error
+        }
+    }
+}
+
+extension Result where Error == Swift.Error {
+    @_transparent
+    public init(_ throwing: () throws -> Value) {
+        do {
+            let value = try throwing()
+            self = .success(value)
+        } catch {
+            self = .failure(error)
+        }
+    }
+    
+    public func unwrap() throws -> Value {
+        switch self {
+        case let .success(value): return value
+        case let .failure(error): throw error
+        }
+    }
+}
+
+extension Result: Equatable where Value: Equatable, Error: Equatable {
+    public static func ==(lhs: Result<Value, Error>, rhs: Result<Value, Error>) -> Bool {
+        return lhs.value == rhs.value && lhs.error == rhs.error
```

Comment:
Since `value` and `error` are both calculated, would it not be more performant to have a switch on lhs and rhs and perform the equality tests there?

Reply:
It'd be good to have a solid answer one way or another.